src/devices/Board/Board.cs

```diff
@@ -0,0 +1,407 @@
+                }
+            }
+        }
+
+        public abstract PinUsage DetermineCurrentPinUsage(int pinNumber);
+
+        /// <summary>
+        /// Override this method if something special needs to be done to use the pin for the given device.
+        /// Many devices support multiple functions per Pin, but not at the same time, so that some kind of
+        /// multiplexer needs to be set accordingly.
+        /// </summary>
+        /// <param name="pinNumber">The logical pin number to use.</param>
+        /// <param name="usage">The intended usage</param>
+        protected virtual void ActivatePinMode(int pinNumber, PinUsage usage)
+        {
+        }
+
+        public virtual void Initialize()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(ToString());
+            }
+
+            _initialized = true;
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            _disposed = true;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private sealed class PinReservation
+        {
+            public PinReservation(int pin, PinUsage usage, object owner)
+            {
+                Pin = pin;
+                Usage = usage;
+                Owner = owner;
+            }
+
+            public int Pin { get; }
+            public PinUsage Usage { get; }
+
+            /// <summary>
+            /// Component that owns the pin (used mainly for debugging)
+            /// </summary>
+            public object Owner { get; }
+        }
+
+        public virtual GpioController CreateGpioController(int[] pinAssignment = null)
+        {
+            return CreateGpioController(pinAssignment, DefaultPinNumberingScheme);
+        }
+
+        public abstract GpioController CreateGpioController(int[] pinAssignment, PinNumberingScheme pinNumberingScheme);
+
+        /// <summary>
+        /// This method shall be overriden by clients, providing just the basic interface to an I2cDevice.
+        /// </summary>
+        /// <param name="connectionSettings">I2C Connection settings</param>
+        /// <param name="pinAssignment">Pins assigned to the device</param>
+        /// <returns>An I2cDevice connection</returns>
+        protected abstract I2cDevice CreateSimpleI2cDevice(I2cConnectionSettings connectionSettings, int[] pinAssignment);
+
+        public I2cDevice CreateI2cDevice(I2cConnectionSettings connectionSettings, int[] pinAssignment, PinNumberingScheme pinNumberingScheme)
+        {
+            if (pinAssignment == null || pinAssignment.Length != 2)
+            {
+                throw new ArgumentException($"Invalid argument. Must provide exactly two pins for I2C", nameof(pinAssignment));
+            }
+
+            return new I2cDeviceManager(this, connectionSettings, RemapPins(pinAssignment, pinNumberingScheme), CreateSimpleI2cDevice);
```

Comment:
tracking connection here with getOrCreate instead of going off with a new operation every time could be interesting

Reply:
That's probably to much hasle. Since you'll not normally use the device directly (but some binding instead), caching seems unnecessary.